src/librustdoc/html/format.rs

```diff
@@ -1085,32 +1085,43 @@ impl Function<'_> {
 }
 
 impl clean::Visibility {
-    crate fn print_with_space<'tcx>(self, tcx: TyCtxt<'tcx>) -> impl fmt::Display + 'tcx {
+    crate fn print_with_space<'tcx>(
+        self,
+        tcx: TyCtxt<'tcx>,
+        item_did: DefId,
+    ) -> impl fmt::Display + 'tcx {
         use rustc_span::symbol::kw;
 
         display_fn(move |f| match self {
             clean::Public => f.write_str("pub "),
             clean::Inherited => Ok(()),
-            clean::Visibility::Restricted(did) if did.index == CRATE_DEF_INDEX => {
-                write!(f, "pub(crate) ")
-            }
-            clean::Visibility::Restricted(did) => {
-                f.write_str("pub(")?;
-                let path = tcx.def_path(did);
-                debug!("path={:?}", path);
-                let first_name =
-                    path.data[0].data.get_opt_name().expect("modules are always named");
-                if path.data.len() != 1 || (first_name != kw::SelfLower && first_name != kw::Super)
-                {
-                    f.write_str("in ")?;
-                }
-                // modified from `resolved_path()` to work with `DefPathData`
-                let last_name = path.data.last().unwrap().data.get_opt_name().unwrap();
-                for seg in &path.data[..path.data.len() - 1] {
-                    write!(f, "{}::", seg.data.get_opt_name().unwrap())?;
+
+            clean::Visibility::Restricted(vis_did) => {
+                if find_closest_parent_module(tcx, item_did) == Some(vis_did) {
+                    // `pub(in foo)` where `foo` is the parent module
+                    // is the same as no visibility modifier
+                    Ok(())
+                } else if vis_did.index == CRATE_DEF_INDEX {
+                    write!(f, "pub(crate) ")
```

Comment:
nit: the `CRATE_DEF_INDEX` check is faster, can you put it first?

Reply:
That changes the behavior (top-level, non-pub items will be displayed as `pub(crate)` instead of no visibility modifier), but I'm doing that anyway.

Reply:
Hmm, the behavior is the same for some reason.

Reply:
I put the check first because we both agree that that behavior is better, but it's not going to be any faster: I extracted a local variable for the parent module because in `pub(super)` case it would compute it twice otherwise. Plus, it's easier to read and less error prone this way. Let me know if you have an idea to get the best of both worlds!